### PR TITLE
feat : add modular_inverse_simple

### PR DESCRIPTION
**Pull request by CarlosZoft**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

this is a simple implementation made during a college course - discrete mathematics

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1937"></a>

### CODE REVIEW

File: math/modular_inverse_simple.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ * @details
+ * this algorithm calculate the modular inverse iteratively
+ *
+ * @ref
+ * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
+ */
+
+#include <cassert>
+#include <iostream>
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:14]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: math/modular_inverse_simple.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @brief Function imod
+ *  Responsible for calculating the modular inverse from x % y
+ * @param x number
+ * @param y number
+ * @returns the modular inverse
+ */
+int imod(int x, int y) {
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:23]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6).** 🙂

File: math/modular_inverse_simple.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    std::cout << "\nPassed!\n";
+
+    std::cout << "\nAll test cases have successfully passed!\n";
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run tests implementations
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:62]:
```suggestion
    test();  // run self-test implementations
    return 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀

File: math/modular_inverse_simple.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#include <iostream>  /// for IO operations
+
+/**
+ * @brief Function imod
+ *  Responsible for calculating the modular inverse from x % y
+ * @param x number
+ * @param y number
+ * @returns the modular inverse
+ */
+uint64_t imod(uint64_t x, uint64_t y) {  // positive integer expected
```
Comment (tjgurwara99) [inline math/modular_inverse_simple.cpp:23]: What happens when a negative number is provided in the arguments - I don't think you've considered a fail safe here.

**Review by tjgurwara99 — DISMISSED** [review]
LGTM

**Comment by CarlosZoft** [comment]
@poyea can you review this pr?

File: math/modular_inverse_simple.cpp (hunk 4 above)
Comment (poyea) [inline math/modular_inverse_simple.cpp:23]: std::uint64_t is defined in header <cstdint>?

Comment (poyea) [inline math/modular_inverse_simple.cpp:23]:
```suggestion
uint64_t imod(uint64_t x, uint64_t y) { 
```

File: math/modular_inverse_simple.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    uint64_t aux = 0;                    // auxiliary variable
+    uint64_t itr = 0;                    // iteration counter
+
+    do {  // run the algorithm while not find the inverse
+        aux = y * itr + 1;
+        itr++;
+    } while (aux % x);  // while module aux % x is different of zero
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:30]:
```suggestion
    } while (aux % x);  // while module aux % x is non-zero
```

File: math/modular_inverse_simple.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/**
+ * @file
+ * @brief simple implementation of adaption modular multiplicative inverse
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:3]:
```suggestion
 * @brief simple implementation of modular multiplicative inverse
```

File: math/modular_inverse_simple.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ * algorithm
+ *
+ * @details
+ * this algorithm calculate the modular inverse iteratively
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:7]:
```suggestion
 * this algorithm calculates the modular inverse iteratively
```

File: math/modular_inverse_simple.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
+/**
+ * @file
+ * @brief simple implementation of adaption modular multiplicative inverse
+ * algorithm
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:4]:
```suggestion
```

File: math/modular_inverse_simple.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ * @ref
+ * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+
+/**
+ * @brief Function imod
+ *  Responsible for calculating the modular inverse from x % y
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:17]: Perhaps
```suggestion
 *  Responsible for calculating the modular inverse of x with respect to y
```

Comment (CarlosZoft) [reply math/modular_inverse_simple.cpp:17]: better, is more descritive

**Comment by poyea** [comment]
@Panquesito7 we need your approval on this

**Review by poyea — DISMISSED** [review]

File: math/modular_inverse_simple.cpp [hunk 10]
```diff
@@ -0,0 +1,62 @@
+/**
+ * @file
+ * @brief simple implementation of modular multiplicative inverse
+ *
+ * @details
+ * this algorithm calculates the modular inverse iteratively
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:6]:
```suggestion
 * this algorithm calculates the modular inverse x^-1 mod y iteratively
```

File: math/modular_inverse_simple.cpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ * @ref
+ * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+
+/**
+ * @brief Function imod
+ *    Responsible for calculating the modular inverse of x with respect to y
```
Comment (poyea) [inline math/modular_inverse_simple.cpp:17]:
```suggestion
 *    Responsible for calculating the modular inverse of x with respect to y, x^{-1} mod y
```

**Review by poyea — DISMISSED** [review]
Thank you for your pull request!🤩

**Comment by CarlosZoft** [comment]
:(

**Comment by poyea** [comment]
@CarlosZoft You need approval from @Panquesito7

File: math/modular_inverse_simple.cpp [hunk 12]
```diff
@@ -0,0 +1,62 @@
+/**
+ * @file
+ * @brief simple implementation of modular multiplicative inverse
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:3]:
```suggestion
 * @brief Simple implementation of [modular multiplicative inverse](https://en.wikipedia.org/wiki/Modular_multiplicative_inverse)
```

File: math/modular_inverse_simple.cpp [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @details
+ * this algorithm calculates the modular inverse x^{-1} \mod y iteratively
+ *
+ * @ref
+ * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:9]:
```suggestion
```

File: math/modular_inverse_simple.cpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    std::cout << "\nPassed!\n";
+
+    std::cout << "\nAll test cases have successfully passed!\n";
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run  self-test implementations
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:61]:
```suggestion
    test();  // run self-test implementations
```

File: math/modular_inverse_simple.cpp [hunk 15]
```diff
@@ -0,0 +1,63 @@
+/**
+ * @file
+ * @brief Simple implementation of [modular multiplicative
+ * inverse](https://en.wikipedia.org/wiki/Modular_multiplicative_inverse)
+ *
+ * @details
+ * this algorithm calculates the modular inverse x^{-1} \mod y iteratively
+ *
+ * @ref
+ * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
```
Comment (Panquesito7) [inline math/modular_inverse_simple.cpp:10]:
```suggestion
```

**Pull request merged**